projects/gsc/experiments/sampled_kwinners/sampled_kwinners.py

```diff
@@ -0,0 +1,246 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+import abc
+
+import numpy as np
+import torch.nn as nn
+
+from nupic.torch.duty_cycle_metrics import max_entropy
+
+from .kwinner_functions import sampled_kwinners, sampled_kwinners2d
+
+
+def update_temperature(m):
+    """Function used to update SampledKWinner modules temperature. This is typically done
+    during training at the beginning of each epoch.
+
+    Call using :meth:`torch.nn.Module.apply` after each epoch if required
+    For example: ``m.apply(update_boost_strength)``
+
+    :param m: KWinner module
+    """
+    if isinstance(m, SampledKWinnersBase):
+        m.update_temperature()
+
+
+class SampledKWinnersBase(nn.Module, metaclass=abc.ABCMeta):
+    """Base SampledKWinners class.
+
+    :param percent_on:
+      k = percent_on * number of input units will be sampled, with replacement, from
+      categorial softmax distribution over the layer activations
+    :type percent_on: float
+
+    :param k_inference_factor:
+      During inference (training=False) we increase percent_on by this factor.
+      percent_on * k_inference_factor must be strictly less than 1.0, ideally much
+      lower than 1.0
+    :type k_inference_factor: float
+
+    :param temperature:
+      temperature to use when passing activations through softmax. Must be >= 1.0
+    :type temperature: float
+
+    :param temperature_decay_rate:
+      reduce the temperature by this much after each epoch, until it hits 1.0.
+    :type temperature_decay_rate: float
+
+    :param eval_temperature:
+      The temperature to use when passing activations through softmax during evaluation.
+      Must be >= 1.0.
+    :type eval_temperature: float
+    """
+
+    def __init__(
+        self,
+        percent_on,
+        k_inference_factor=1.0,
+        temperature=10.0,
+        temperature_decay_rate=0.01,
+        eval_temperature=1.0,
+    ):
+        super(SampledKWinnersBase, self).__init__()
+        assert 0.0 < percent_on < 1.0
+        assert 0.0 < percent_on * k_inference_factor < 1.0
+        assert eval_temperature >= 1.0
+
+        self.percent_on = percent_on
+        self.percent_on_inference = percent_on * k_inference_factor
+        self.temperature = temperature
+        self.temperature_decay_rate = temperature_decay_rate
+        self.eval_temperature = eval_temperature
+        self.learning_iterations = 0
+        self.n = 0
+        self.k = 0
+        self.k_inference = 0
+
+    def _load_from_state_dict(self, *args, **kwargs):
```

Comment:
I don't think there's a need to define this. Nothing is added on line 96. Unless there's something I'm missing, you can remove it.

Reply:
is loading from a state dictionary a functionality that we don't want?

Reply:
It is, but I don't believe you need to define `_load_from_state_dict` to have the functionality. The parent class, `torch.nn.Module`, already has it. In your overridden definition here, you're simply calling the parent's implementation via `super()` and nothing else is being added. Hence, you can remove yours here.